Order Matching

amustapha opened the pull request:
- Ensures cancelling order restores balances ✅
- Implement order exectuion and ensure balances are accurate afterwards ✅
- Ensure orders are executed on each tick ✅
- Handle post only limit orders ✅ 
- Implement fetch_my_trades ✅
- Enhance testing for fetch closed trades ✅
-

amustapha committed:
feat: improve order cancellation logic to restore asset balances

amustapha committed:
feat: enhance create_order method to handle immediate fill scenarios for limit orders

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `97%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `88%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `100%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
test: move backtester fixture with data feeds to conftest for reusability

amustapha committed:
fix: correct price assignment for market orders in backtester logic

amustapha committed:
fix: fix if..elif logic to be explicit

amustapha committed:
test: add tests for limit order behavior with postOnly parameter

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `98%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `100%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
fix: clarify conditional logic for sell side in backtester

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `98%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `100%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
test: scaffold tests for balance modification post order execution

amustapha committed:
feat: implement order filling logic and update balance accordingly

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `98%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `95%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
feat: add fetch_my_trades method and ensure orders fill correctly on each tick

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `98%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `95%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
test: add error handling test for adding existing data feed in backtester

amustapha commented:
| File | Coverage |   |
| - | :-: | :-: |
| **All files** | `98%` | :white_check_mark: |
| ccxt_backtesting_exchange/__init__.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | :white_check_mark: |
| ccxt_backtesting_exchange/clock.py | `100%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | :white_check_mark: |
| ccxt_backtesting_exchange/utils.py | `100%` | :white_check_mark: |
| tests/__init__.py | `100%` | :white_check_mark: |
| tests/conftest.py | `100%` | :white_check_mark: |
| tests/test_backtester_balance.py | `100%` | :white_check_mark: |
| tests/test_backtester_market.py | `100%` | :white_check_mark: |
| tests/test_backtester_orders.py | `100%` | :white_check_mark: |
| tests/test_clock.py | `100%` | :white_check_mark: |
| tests/test_data_feed.py | `100%` | :white_check_mark: |
| tests/test_utils_timeframe_parser.py | `100%` | :white_check_mark: |
| tests/utils.py | `95%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

amustapha committed:
chore: update test coverage workflow to  enhance reporting

amustapha commented:
| File | Coverage | Lines | Branches |   |
| - | :-: | :-: | :-: | :-: |
| **All files** | `98%` | `98%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `94%` | `94%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | `99%` | `0%` | :white_check_mark: |
| tests/utils.py | `95%` | `95%` | `0%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

peterolayinka committed:
Merge remote-tracking branch 'origin/master' into feat/order-matching

amustapha commented:
| File | Coverage | Lines | Branches |   |
| - | :-: | :-: | :-: | :-: |
| **All files** | `98%` | `98%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `95%` | `95%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | `99%` | `0%` | :white_check_mark: |
| tests/utils.py | `95%` | `95%` | `0%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_

peterolayinka commented on ccxt_backtesting_exchange/backtester.py:
```diff
@@ -223,12 +231,60 @@ def _update_asset_balance(self, asset: str, column: str, amount: float) -> None:
+            base_asset, quote_asset = symbol.split("/")
+            for index, order in open_orders[open_orders["symbol"] == symbol].iterrows():
+                trade_value = order["amount"] * order["price"]
+                if order["price"] >= low and order["price"] <= high:
```
@amustapha Should we not be using the open/close price instead of checking if the order was within the range? On a second though using the range seems a bit more valid because that shows the order must have executed within that timeframe, though we can't tell when exactly. I guess this should be fine.

amustapha replied:
exactly. the only con is, if price is exactly low, we might have a patial fill and that's hard to predetermine. using > or < explicitly would also be inaccurate.

peterolayinka committed:
Confirm that the order were actually closed

If the order isn't open, it could be cancelled, I guess it makes sense to be explicit about it status.

peterolayinka reviewed (approved):
Wizardry at it best. I can confirm the logic and test all worked as expected.

amustapha commented:
| File | Coverage | Lines | Branches |   |
| - | :-: | :-: | :-: | :-: |
| **All files** | `98%` | `98%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/backtester.py | `95%` | `95%` | `0%` | :white_check_mark: |
| ccxt_backtesting_exchange/data_feed.py | `99%` | `99%` | `0%` | :white_check_mark: |
| tests/utils.py | `95%` | `95%` | `0%` | :white_check_mark: |

_Minimum allowed coverage is `75%`_
